# **Homework 1: Data Munging and Subsetting (30 points)**

#### **Dylan Watson**


---
**Instructions**: Complete the homework assignment below.
- <font color=red>Tasks</font> indicated in red must be completed to receive credit
- Include all code for your work
- Include markdown cells with written answers explaining your work when prompted
- Your completed assignment must be submitted to Canvas as a .ipynb file by the deadline listed on Canvas
---

####<font color=red>Organized</font> (1 points)
* Answers should be organized in the following format for organization and readability

<br>

```
# 1. 
This will be the code block to answer the first part of a task
```

1. This will be the text answer/explanation (if prompted) for the first task

<br>

```
# 2. 
This will be the code block to answer the second part of a task in a separate code cell
```

2. This will be the text answer/explanation (if prompted) for the second task

<br>

Continue this format for the remaining subtasks


<br>

---

<br>






####<font color=red>Task 0</font>: Subsetting Practice  (10 points)

Using subsetting to perform the following tasks using the `mtcars` dataset
1. Select all rows of cars corresponding to 6 cylinder engines AND 3 gears

2. Select first 2 rows and the 3rd and 5th columns of the `mtcars` dataset using subsetting

3. Select all rows corresponding to horsepower (`hp`) greater than 300

4. Using subsetting to calculate the average weight `wt` of cars with `mpg` < 15

5. Using subsetting to calculate the average weight `wt` of cars with `mpg` < 15 OR `mpg` > 30


In [1]:
data(mtcars)
# subset 1
mtcars[mtcars$cyl == 6 & mtcars$gear == 3, ]
# subset 2
mtcars[c(1:2), c(3, 5)]
# subset 3
mtcars[mtcars$hp > 300, ]
# subset 4
mean(mtcars$wt[mtcars$mpg < 15])
# subset 5
mean(mtcars$wt[mtcars$mpg < 15 | mtcars$mpg > 30])

mpg  cyl disp hp  drat wt    qsec  vs am gear carb
Hornet 4 Drive 21.4 6   258  110 3.08 3.215 19.44 1  0  3    1   
Valiant        18.1 6   225  105 2.76 3.460 20.22 1  0  3    1

disp drat
Mazda RX4     160  3.9 
Mazda RX4 Wag 160  3.9

mpg cyl disp hp  drat wt   qsec vs am gear carb
Maserati Bora 15  8   301  335 3.54 3.57 14.6 0  1  5    8

[1] 4.6858

[1] 3.399111

<br>

---

<br>






####<font color=red>Task 1</font>  (1 point)

1. Load the `nst-est2011-06.csv` from the [census website](https://www2.census.gov/programs-surveys/popest/tables/2010-2011/state/totals/)
2. Store the dataset as a dataframe called `events`


In [2]:
events = read.csv('nst-est2011-06.csv')
head(events, 8)
tail(events, 12)

table.with.row.headers.in.column.A.and.column.headers.in.rows.3.through.4...leading.dots.indicate.sub.parts.                                                                  
1 Table 6. Estimates of the Annual Rates1 of the Components of Resident Population Change for the United States,\nRegions, States, and Puerto Rico: July 1, 2010 to July 1, 2011
2 Geographic Area                                                                                                                                                               
3                                                                                                                                                                               
4 United States                                                                                                                                                                 
5 Northeast                                                                                                                                                                     
6 Midwest                                                                                                                                                                       
7 South                                                                                                                                                                         
8 West                                                                                                                                                                          
  X                        X.1               X.2          X.3    X.4          
1                                                                             
2 Total Population Change2 Natural\nIncrease Vital Events        Net Migration
3                                            Births       Deaths Total        
4 7.3                      5.0               12.9         7.9    2.3          
5 2.8                      3.4               11.6         8.3    -0.5         
6 2.7                      4.1               12.6         8.5    -1.4         
7 10.3                     5.1               13.3         8.2    5.2          
8 10.1                     7.0               13.6         6.6    3.2          
  X.5            X.6      X.7 X.8 X.9 X.10
1                         NA  NA  NA  NA  
2                         NA  NA  NA  NA  
3 International3 Domestic NA  NA  NA  NA  
4 2.3            (X)      NA  NA  NA  NA  
5 2.6            -3.1     NA  NA  NA  NA  
6 1.3            -2.7     NA  NA  NA  NA  
7 2.3            2.9      NA  NA  NA  NA  
8 2.9            0.2      NA  NA  NA  NA

table.with.row.headers.in.column.A.and.column.headers.in.rows.3.through.4...leading.dots.indicate.sub.parts.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            
60                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
61 Puerto Rico                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
62 1 Rates per 1,000 average population.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
63 2 Total population change includes a residual.  This residual represents the change in population that cannot be attributed to any specific demographic component.  See State and County Terms and Definitions at http://www.census.gov/popest/about/terms/stco.html.                                                                                                                                                                                                                                                                                                                                                                                   
64 3 Net international migration for the United States, regions, and states includes the international migration of both native and foreign-born populations. Specifically, it includes: (a) the net international migration of the foreign born, (b) the net migration between the United States and Puerto Rico, (c) the net migration of natives to and from the United States, and (d) the net movement of the Armed Forces population between the United States and overseas.  Net international migration for Puerto Rico includes only the international migration of native and foreign-born populations between the United States and Puerto Rico.
65 Dash (-) represents zero or rounds to zero.                                                                                                                                             

<br>

---

<br>






####<font color=red>Task 2</font>  (2 points)

1. Use the data munging techniques from lecture to create a dataframe called `states`
  * The dataframe should have a row for every state:
    * Exclude Puerto Rico
    * You may include Washington DC
  * The dataframe should have 8 columns
    * state
    * population change
    * natural increase
    * births
    * deaths
    * migration total
    * international migration
    * domestic migration
2. Show the `head()` and `tail()` of the dataframe `states` to confirm the data is formatted properly


In [3]:
# Subset data
nb_row = nrow(events)
states <- events[-c(1:8, (nb_row - 11):nb_row), -(9:12)]

# Change column names
names(states) <- c("Geographic Area", "Total Population Change", "Natural Increase", "Vital Events: Births", "Vital Events: Deaths", "Net Migration: Total", "Net Migration: International", "Net Migration: Domestic")

# Remove .
states$'Geographic Area' <- gsub('\\.', '', states$'Geographic Area')

# Convert to numeric and remove -
states$'Net Migration: Domestic' <- as.numeric(gsub("-", "0", states$'Net Migration: Domestic'))
states[, 2:7] <- apply(states[, 2:7], 2, function(x) as.numeric(as.character(x)))

rownames(states) <- NULL

# Check
head(states)
tail(states)

Geographic Area Total Population Change Natural Increase Vital Events: Births
1 Alabama          3.6                     2.5             12.5                
2 Alaska          11.9                    10.9             15.6                
3 Arizona         10.8                     6.5             13.6                
4 Arkansas         5.6                     3.3             13.0                
5 California       9.4                     7.3             13.5                
6 Colorado        13.6                     6.9             13.0                
  Vital Events: Deaths Net Migration: Total Net Migration: International
1 10.0                 1.1                  0.9                         
2  4.6                 1.0                  1.5                         
3  7.1                 4.3                  3.2                         
4  9.7                 2.3                  1.1                         
5  6.2                 2.1                  3.5                         
6  6.1                 6.6                  1.7                         
  Net Migration: Domestic
1 0.1                    
2 0.6                    
3 1.1                    
4 1.2                    
5 1.4                    
6 4.9

Geographic Area Total Population Change Natural Increase
46 Vermont          0.8                     1.4            
47 Virginia         9.0                     5.3            
48 Washington      12.8                     5.6            
49 West Virginia    0.5                    -0.5            
50 Wisconsin        3.5                     4.1            
51 Wyoming          6.4                     5.5            
   Vital Events: Births Vital Events: Deaths Net Migration: Total
46  9.5                  8.1                 -0.5                
47 12.6                  7.3                  3.7                
48 12.8                  7.1                  7.2                
49 11.1                 11.7                  1.1                
50 12.2                  8.1                 -0.5                
51 13.3                  7.8                  0.9                
   Net Migration: International Net Migration: Domestic
46 0.6                          1.1                    
47 2.4                          1.3                    
48 2.5                          4.6                    
49 0.4                          0.7                    
50 0.9                          1.4                    
51 0.8                          0.0

<br>

---

<br>






####<font color=red>Task 3</font>  (2 points)

Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.

1. Which state has the lowest deaths per 1,000 people?
2. Which state has the highest deaths per 1,000 people?

In [4]:
states[which.min(states$'Vital Events: Deaths'), c(1, 5)]
states[which.max(states$'Vital Events: Deaths'), c(1, 5)]

Geographic Area Vital Events: Deaths
2 Alaska          4.6

Geographic Area Vital Events: Deaths
49 West Virginia   11.7

The state with the lowest deaths in the June 2011 census was Alaska at a rate of 4.6 per 1,000 people. The state with the highest deaths in the June 2011 census was West Virginia at a rate of 4.6 per 1,000 people.

<br>

---

<br>






####<font color=red>Task 4</font>  (2 points)

Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.

1. Which state has the lowest births per 1,000 people?
2. Which state has the highest births per 1,000 people?

In [5]:
states[which.min(states$'Vital Events: Births'), c(1, 4)]
states[which.max(states$'Vital Events: Births'), c(1, 4)]

Geographic Area Vital Events: Births
46 Vermont         9.5

Geographic Area Vital Events: Births
45 Utah            18.6

The state with the lowest births is Vermont at 9.5 per 1,000 people. The state with the greatest births is Utah at 18.6 per 1,000 people. 

<br>

---

<br>






####<font color=red>Task 5</font>  (2 points)

Think about the results you observe from Tasks 3 and 4. 
* Are these surpising? Are these expected? Why?

These results from tasks 3 and 4 are not surprising. Alaksa has a low death rate because it has a significantly younger avg population than most states. Vermont has a significantly older population thus it has less young families bearing children. Utah has a large mormon population which believes in significantly larger families for reasons of faith that result in a high birth rate. 

<br>

---

<br>






####<font color=red>Task 6</font>  (3 points)

Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.

1. Which states have more deaths than births per 1,000 people?
2. Which states have a negative population change?
3. Which states have a population change greater than 15 OR less than 0?

In [6]:
states[states$'Vital Events: Deaths' > states$'Vital Events: Births', c(1, 4:5)]
states[states$'Total Population Change' < 0, c(1:2)]
states[states$'Total Population Change' > 15 | states$'Total Population Change' < 0, c(1:2)]

Geographic Area Vital Events: Births Vital Events: Deaths
49 West Virginia   11.1                 11.7

Geographic Area Total Population Change
23 Michigan        -0.1                   
40 Rhode Island    -1.2

Geographic Area      Total Population Change
9  District of Columbia 21.4                   
23 Michigan             -0.1                   
40 Rhode Island         -1.2                   
44 Texas                16.5

The only state to have a higher death than birth rate is West Virginia. The states with negative population change are Michigan and Rhode Island. The states with a population change greater than 15 or less than 0 are DC, Michigan, Rhode Island, and Texas.

<br>

---

<br>






####<font color=red>Task 7</font>  (3 points)

Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.

1. What is the average number of births across states with population changes greater than 10 per 1,000 people.
2. What is the average number of births across states with population changes less than 5 per 1,000 people.
3. Are these averages very different?

In [7]:
states[states$'Total Population Change' > 10, c(1:2,4)]
mean(states$'Vital Events: Births'[states$'Total Population Change' > 10])
states[states$'Total Population Change' < 5, c(1:2,4)]
mean(states$'Vital Events: Births'[states$'Total Population Change' < 5])

Geographic Area      Total Population Change Vital Events: Births
2  Alaska               11.9                    15.6                
3  Arizona              10.8                    13.6                
6  Colorado             13.6                    13.0                
9  District of Columbia 21.4                    14.7                
10 Florida              11.6                    11.3                
11 Georgia              10.6                    13.8                
35 North Dakota         13.7                    13.2                
44 Texas                16.5                    15.6                
45 Utah                 14.9                    18.6                
48 Washington           12.8                    12.8

[1] 14.22

Geographic Area Total Population Change Vital Events: Births
1  Alabama          3.6                    12.5                
7  Connecticut      1.5                    10.4                
14 Illinois         2.1                    13.0                
15 Indiana          4.0                    13.0                
16 Iowa             4.0                    12.5                
17 Kansas           4.2                    14.0                
20 Maine            0.6                     9.7                
22 Massachusetts    4.9                    11.2                
23 Michigan        -0.1                    11.5                
25 Mississippi      2.8                    13.6                
26 Missouri         2.5                    12.6                
30 New Hampshire    1.1                     9.7                
31 New Jersey       2.4                    11.8                
33 New York         3.6                    12.5                
36 Ohio             0.6                    12.2                
39 Pennsylvania     2.0                    11.3                
40 Rhode Island    -1.2                    10.5                
46 Vermont          0.8                     9.5                
49 West Virginia    0.5                    11.1                
50 Wisconsin        3.5                    12.2

[1] 11.74

For states with population change greater than 10, the avg births is 14.22 per 1,000. For states with pop change less than 5, the avg births is 11.74 per 1,000. Yes there is a significant difference in birthrates between states with greater and less pop changes. 

<br>

---

<br>






####<font color=red>Task 8</font>  (4 points)

Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.

1. Order the dataframe from lowest deaths per 1,000 to highest deaths per 1,000
2. Store the dataframe as `states_sorted`
3. Show the first few rows to ensure they are sorted correctly
4. Which 5 states have the highest number of deaths per 1,000?

In [8]:
states_sorted <- states[order(states$'Vital Events: Deaths'), ]
head(states_sorted)
tail(states_sorted)

Geographic Area Total Population Change Natural Increase
2  Alaska          11.9                    10.9            
45 Utah            14.9                    13.5            
6  Colorado        13.6                     6.9            
5  California       9.4                     7.3            
44 Texas           16.5                     9.1            
3  Arizona         10.8                     6.5            
   Vital Events: Births Vital Events: Deaths Net Migration: Total
2  15.6                 4.6                  1.0                 
45 18.6                 5.1                  1.5                 
6  13.0                 6.1                  6.6                 
5  13.5                 6.2                  2.1                 
44 15.6                 6.5                  7.4                 
3  13.6                 7.1                  4.3                 
   Net Migration: International Net Migration: Domestic
2  1.5                          0.6                    
45 1.9                          0.4                    
6  1.7                          4.9                    
5  3.5                          1.4                    
44 2.9                          4.5                    
3  3.2                          1.1

Geographic Area Total Population Change Natural Increase
20 Maine           0.6                      0.1            
4  Arkansas        5.6                      3.3            
25 Mississippi     2.8                      3.9            
39 Pennsylvania    2.0                      1.4            
1  Alabama         3.6                      2.5            
49 West Virginia   0.5                     -0.5            
   Vital Events: Births Vital Events: Deaths Net Migration: Total
20  9.7                  9.6                  0.5                
4  13.0                  9.7                  2.3                
25 13.6                  9.7                 -1.1                
39 11.3                  9.8                  0.6                
1  12.5                 10.0                  1.1                
49 11.1                 11.7                  1.1                
   Net Migration: International Net Migration: Domestic
20 0.4                          0.1                    
4  1.1                          1.2                    
25 0.8                          1.9                    
39 1.2                          0.6                    
1  0.9                          0.1                    
49 0.4                          0.7

The top 5 states by death rate are West Virginia, Alabama, Pennsylvania, Mississippi, and Arkansas.

<br>

---

<br>




